### PR TITLE
Fix comments' position

### LINKED ISSUE #917: comments moving in ways that change their meaning
I am using ocamlformat built from 2c8d4544d794b6ecd8645b11b3a071fd5f04cf83.

This program ...
```
let _ =
  try_with (fun () ->
    match get () with
    | None -> do_something ()
    | Some _ -> () (* do nothing *))
;;
```
... gets rewritten to this.
```
let _ =
  try_with (fun () ->
    (* do nothing *)
    match get () with
    | None -> do_something ()
    | Some _ -> ())
;;
```

This program ...
```
let _ =
  try_with (fun () ->
    a;
    b (* after a *))
;;
```
... gets rewritten to this.
```
let _ =
  try_with (fun () ->
    (* after a *)
    a;
    b)
;;
```

In both cases the comment becomes nonsensical.

### PR DESCRIPTION
**Pull request by hhugo**
attempt to fix #917

### CODE REVIEW

File: src/Ast.ml [hunk 1]
```diff
@@ -2216,7 +2216,9 @@ end = struct
       in
       match ambig_prec (sub_ast ~ctx (Exp exp)) with
       | None -> false (* ctx not apply *)
+
```
Comment (hhugo) [inline src/Ast.ml]: @gpetiot, I don't understand why we have this newline here..

Comment (gpetiot) [reply src/Ast.ml]: I'm looking into this, but first I'm trying to fix the other comments-dropped issues that arise after the previous commits of this PR.

Comment (gpetiot) [reply src/Ast.ml]: This is fixed, but there are still some other regressions to fix in the testsuite.

**Comment by gpetiot** [comment]
Last regression that I didn't manage to fix yet (with `--margin=100`), this doesn't stabilize after 10 iterations:
```ocaml
let prop_init_formals_seed fooooooooooooooooooo =
  let sigma_seed =
    create_seed_vars (* formals already there plus new ones *) (prop.Prop.sigma @ sigma_new_formals)
  in
  foooooooooooooooooooooooooo
```

File: test/passing/comments.ml.ref [hunk 2]
```diff
@@ -1,4 +1,4 @@
-let _ = f (*f*) a (*a*) ~b (*comment*) ~c:(*comment*) c' ?d ?e ()
+let _ = f (*f*)  a (*a*)  ~b (*comment*) ~c:(*comment*) c' ?d ?e ()
```
Comment (gpetiot) [inline test/passing/comments.ml.ref]: These spaces should be removed (and more below).

**Comment by gpetiot** [comment]
@hhugo I fixed the last regressions I think it's ready to be reviewed now.

File: test/passing/js_source.ml.ref [hunk 3]
```diff
@@ -522,7 +521,7 @@ type _ foo += C : ('a -> int) -> 'a foo
 (* ERROR: Parameter variances are not satisfied *)
 
 type 'a bar = ..
-type +'a bar += D of (int -> 'a) (* ERROR: type variances do not match *)
```
Comment (hhugo) [inline test/passing/js_source.ml.ref]: the comment moved inside the parens.

File: test/passing/js_source.ml.ref [hunk 4]
```diff
@@ -2339,11 +2336,9 @@ let f : type env a. (env, a) typ -> (env, a) typ -> int =
   | Tint, Tint -> 0
   | Tbool, Tbool -> 1
   | Tvar var, tb -> 2
-  | _ -> .
```
Comment (hhugo) [inline test/passing/js_source.ml.ref]: This is great !

File: test/passing/js_source.ml.ref [hunk 5]
```diff
@@ -2506,11 +2501,9 @@ let example6 : type a. a wrapPoly -> a -> int =
  fun w ->
   match w with
   | WrapPoly ATag -> intA
-  | WrapPoly _ -> intA
```
Comment (hhugo) [inline test/passing/js_source.ml.ref]: This is great !

File: test/passing/js_source.ml.ref [hunk 6]
```diff
@@ -5082,12 +5071,11 @@ module type S = sig
 end
 
 let f (type a) (module X : S with type t = a) = ()
-let _ = f (module A) (* ok *)
```
Comment (hhugo) [inline test/passing/js_source.ml.ref]: moved inside parens

File: test/passing/js_source.ml.ref [hunk 7]
```diff
@@ -5082,12 +5071,11 @@ module type S = sig
...
+let _ = f (module A (* ok *))
 
 module A_annotated_alias : S with type t = (module A.A_S) = A
 
-let _ = f (module A_annotated_alias) (* ok *)
```
Comment (hhugo) [inline test/passing/js_source.ml.ref]: moved inside parens

File: test/passing/js_source.ml.ref [hunk 8]
```diff
@@ -5097,12 +5085,9 @@ module A_alias_expanded = struct
 end
 
 let _ = f (module A_alias_expanded : S with type t = (module A.A_S)) (* ok *)
-
```
Comment (hhugo) [inline test/passing/js_source.ml.ref]: moved inside parens

File: test/passing/label_option_default_args.ml.ref [hunk 9]
```diff
@@ -4,104 +4,66 @@ let (* 0 *) f (* 1 *) x (* 2 *) = (* 3 *) e
 
 let f ~x = e
 
-let (* 0 *) f ~(* 1 *) x (* 2 *) = (* 3 *) e
```
Comment (hhugo) [inline test/passing/label_option_default_args.ml.ref]: This is great !

File: test/passing/open.ml.ref [hunk 10]
```diff
@@ -167,13 +167,15 @@ let x =
...
   (* A'' *)
-  let open (* A' *) A (* A *) in
+  let open A (* A *) in
+  (* A' *)
   (* B *)
   b
 
 class a =
   (* A'' *)
-  let open (* A' *) A (* A *) in
```
Comment (hhugo) [inline test/passing/open.ml.ref]: This is weird

File: test/passing/open.ml.ref [hunk 11]
```diff
@@ -182,23 +184,24 @@ let _ =
...
-  end
-  (* d *) in
+
+    (* d *)
+  end in
   (* e *)
   (* f *)
   let (* g *)
-  open (* h *) A (* i *) (B)
-  (* j *) in
```
Comment (hhugo) [inline test/passing/open.ml.ref]: `(* j *)` moved inside.

File: test/passing/open_preserve.ml.ref [hunk 12]
```diff
@@ -182,23 +184,24 @@ let _ =
   let (* b *)
   open (* c *) struct
     type t
-  end
-  (* d *) in
```
Comment (hhugo) [inline test/passing/open_preserve.ml.ref]: comment move inside.

File: test/passing/open.ml.ref (hunk 10 above)
Comment (gpetiot) [reply test/passing/open.ml.ref]: I didn't manage to fix this one, nothing seems wrong in the code.

**Comment by gpetiot** [comment]
I optimized a bit the changes made in Cmts.ml, but it can still be improved, any idea is welcomed.

**Comment by hhugo** [comment]
I believe I've fixed the performance issue.

File: test/passing/js_source.ml.ref [hunk 13]
```diff
@@ -8390,11 +8375,15 @@ end
...
 
-  type state (* the state of a Lua interpreter *)
+  (* a Lua value *)
+  type state
+
+  (* the state of a Lua interpreter *)
+  type usert
 
-  type usert (* a user-defined value *)
+  (* a user-defined value *)
```
Comment (Julow) [inline test/passing/js_source.ml.ref]: These 3 comments are not placed correctly, the original code is:

```ocaml
module type VALUE = sig
  type value (* a Lua value *)
  type state (* the state of a Lua interpreter *)
  type usert (* a user-defined value *)
end;;
```

File: test/passing/open_preserve.ml.ref (hunk 12 above)
Comment (gpetiot) [reply test/passing/open_preserve.ml.ref]: This is due to the new within/within heuristics, and I didn't manage to fix this without reverting to the state previous to this PR.

**Comment by hhugo** [comment]
> This PR is very big. Is there parts that could be moved to new PRs to make review easier ?

The diff in Translation_unit can be moved in a separate PR

**Comment by gpetiot** [comment]
Should revert #923 once it is merged.

**Comment by hhugo** [comment]
rebased

**Comment by Julow** [comment]
This PR fixes https://github.com/ocaml-ppx/ocamlformat/issues/989

File: test/passing/js_source.ml.ref (hunk 13 above)
Comment (Julow) [reply test/passing/js_source.ml.ref]: This bug is still here.

File: test/passing/open-auto.ml.ref [hunk 14]
```diff
@@ -164,13 +164,15 @@ let x =
 
 class type a =
   (* A'' *)
-  let open (* A' *) A (* A *) in
+  let open A (* A *) in
+  (* A' *)
```
Comment (Julow) [inline test/passing/open-auto.ml.ref]: This comment moved.

File: src/Fmt_ast.ml [hunk 15]
```diff
@@ -894,6 +894,113 @@ and fmt_pattern c ?pro ?parens ({ctx= ctx0; ast= pat} as xpat) =
   update_config_maybe_disabled c ppat_loc ppat_attributes
   @@ fun c ->
   let parens = match parens with Some b -> b | None -> parenze_pat xpat in
+  let width xpat = String.length (Cmts.preserve (fmt_pattern c) xpat) in
+  let fmt_op_args op_args =
```
Comment (Julow) [inline src/Fmt_ast.ml]: This function is very similar with the one in `fmt_expression`. Would it be possible to share some parts ? (eg. the ones that implement formatting options)

Comment (gpetiot) [reply src/Fmt_ast.ml]: Should be factorized with the new `fmt_infix_op_args` option

**Comment by gpetiot** [comment]
Fixed the conflicts but it is worse than before, more work needed to the attachment algorithm.

**Comment by Julow** [comment]
What should we do with this PR ?

**Comment by gpetiot** [comment]
Hopefully #1193 will fix everything this issue ties to fix.

**Comment by jberdine** [comment]
Should this be closed or converted to a draft?

**Pull request closed without merging**